### PR TITLE
Test refactor

### PR DESCRIPTION
**Pull request by fabianvf**
add module that generates test classes for each resource that has defined examples

Would allow you to add tests for a resource by just writing the docstring example for it.

I wanted to parametrize each of the test functions by the `{create|patch|remove|replace}_params` but pytest doesn't currently support parametrizing with a fixture.

### CODE REVIEW

**Comment by fabianvf** [comment]
ping @chouseknecht

File: test/functional/conftest.py [hunk 1]
```diff
@@ -277,35 +249,17 @@ def _get_id(argvalue):
...
-    if 'replace_tasks' in metafunc.fixturenames:
-        tasks = [x for x in data if x.get('replace')]
-        metafunc.parametrize("replace_tasks", tasks, False, _get_id)
-    if 'namespaces' in metafunc.fixturenames:
-        tasks = [x for x in data if x.get('create') and x['create'].get('namespace')]
-        unique_namespaces = dict()
-        for task in tasks:
-            unique_namespaces[task['create']['namespace']] = None
-        metafunc.parametrize("namespaces", list(unique_namespaces.keys()))
+# def pytest_generate_tests(metafunc):
```
Comment (detiber) [inline test/functional/conftest.py]: Can you kill these comments?

**Comment by fabianvf** [comment]
@detiber I think the number of tests being run is because I'm not currently generating one test per example listed, I'm generating one test for create/patch/replace/get/delete (or 5 tests per resource), which are then run if there are examples for the given verb. So if you have 10 service create examples, it will just show up as 1 test.

Personally I hate that, but pytest has really bad support for parameterizing tests based on fixtures (what I'd like to do is have the test_{create|get|patch|replace|delete} parameterized by the incoming tasks). It's also super awkward with replace/patch becuase you get the combinatorial explosion of `create_tasksX{patch|replace}_tasks` because of the resources fixture.

**Comment by detiber** [comment]
> @detiber I think the number of tests being run decreasing is because I'm not currently generating one test per example listed, I'm generating one test for create/patch/replace/get/delete (or 5 tests per resource), which are then run if there are examples for the given verb. So if you have 10 service create examples, it will just show up as 1 test.

Does this mean that multiple replace/patch/delete examples would be run against the same fixture? I'm wondering if that would lead to potential issues.

The other thing that I wonder about is how would we provide tests that we expect to fail? For example trying to mutate an immutable field?

**Comment by fabianvf** [comment]
> Does this mean that multiple replace/patch/delete examples would be run against the same fixture? I'm wondering if that would lead to potential issues.

There definitely is some complexity there, though I think it existed in the previous iteration as well (I don't think all patch tasks can be applied to all create tasks). Best way I've thought of to handle this is to add some metadata to the resource example that ties it to a specific create task. Then, once pytest gets their fixture parameterization working (or before that, just a lot more work), we can parameterize the test with pairs of create/{patch|replace} tasks that we are testing.

> The other thing that I wonder about is how would we provide tests that we expect to fail? For example trying to mutate an immutable field?

This is also something we could handle by extending the example metadata, and just having logic to check if the test is supposed to fail or not and handling the logic for it.

Really my biggest want right now is to get each test running one task, that would make both of the above issues much easier to handle. Not sure if you have any thoughts on how to accomplish that, I  spun my wheels on it for a while but didn't come up with anything.

**Comment by fabianvf** [comment]
It also seems like the tests just broke, likely as a result of issues with patching the resources.

**Comment by detiber** [comment]
@fabianvf the tests seem to be flaky in general right now, they are failing against master as well. Haven't had a chance to dig in more deeply yet.

**Comment by fabianvf** [comment]
@detiber interesting, yeah it seems they are failing with the same tests in the same versions as well. Bizarre, was there an update to the openshift containers those versions are using?

**Comment by fabianvf** [comment]
Some documentation of pytest issues; https://github.com/pytest-dev/pytest/issues/349

**Comment by detiber** [comment]
@fabianvf PR to your branch incoming related to test failures, the test still fails for the latest image, which actually should be expected because those fields are now immutable, but it turns out that due to a variable name typo the 1.5 tests were running against the latest image.

**Comment by fabianvf** [comment]
Failing test cases are legitimate failures.

**Pull request merged**